internal/transport/http2_client.go

```diff
@@ -1006,24 +1008,30 @@ func (t *http2Client) Close(err error) {
 		t.kpDormancyCond.Signal()
 	}
 	t.mu.Unlock()
+	var st *status.Status
 	// Per HTTP/2 spec, a GOAWAY frame must be sent before closing the
-	// connection. See https://httpwg.org/specs/rfc7540.html#GOAWAY.
+	// connection. See https://httpwg.org/specs/rfc7540.html#GOAWAY. It
+	// also waits for loopyWriter to be closed with a timer to avoid the
+	// long blocking in case the connection is half closed.
 	t.controlBuf.put(&goAway{code: http2.ErrCodeNo, debugData: []byte("client transport shutdown"), closeConn: err})
-	<-t.writerDone
+	select {
+	case <-t.writerDone:
+		// Append info about previous goaway's if there were any, since this
+		// may be important for understanding the root cause for this
+		// connection to be closed.
+		_, goAwayDebugMessage := t.GetGoAwayReason()
+		if len(goAwayDebugMessage) > 0 {
+			st = status.Newf(codes.Unavailable, "closing transport due to: %v, received prior goaway: %v", err, goAwayDebugMessage)
+			err = st.Err()
+		} else {
+			st = status.New(codes.Unavailable, err.Error())
+		}
+	case <-time.After(goAwayLoopyWriterTimeout):
+		t.logger.Warningf("Failed to write a GOAWAY frame as part of connection close after %v. Giving up and closing the transport.", goAwayLoopyWriterTimeout)
```

Comment:
I believe no as in else case, we wouldn't had sent the GOAWAY as we failed to do so. WDYT?

Reply:
I think you still have to set the status code because you are closing the transport. Probably similar to the else part of writer done i.e `code.Unavailable`?